docs/architecture-overview.rst

```diff
@@ -0,0 +1,200 @@
+; integrations with `Phabricator
+<https://github.com/zulip/phabricator-to-zulip>`__, `Jenkins
+<https://github.com/zulip/zulip-jenkins-plugin>`__, `Puppet
+<https://github.com/matthewbarr/puppet-zulip>`__, `Redmine
+<https://github.com/zulip/zulip-redmine-plugin>`__, and `Trello
+<https://github.com/zulip/trello-to-zulip>`__); `node.js API bindings
+<https://github.com/zulip/zulip-node>`__; and our `full-text search
+PostgreSQL extension <https://github.com/zulip/tsearch_extras>`__ .
+
+Our mobile clients are separate code repositories: `desktop
+<https://github.com/zulip/zulip-desktop>`__, `Android
+<https://github.com/zulip/zulip-android>`__, `iOS (stable)
+<https://github.com/zulip/zulip-ios>`__ , and `our experimental React
+Native iOS app <https://github.com/zulip/zulip-mobile>`__.
+
+We use `Transifex <https://www.transifex.com/zulip/zulip/>`__ to do
+translations.
+
+In this overview we'll mainly discuss the core Zulip web application.
+
+
+Usage assumptions and concepts
+==============================
+
+Zulip is a real-time web-based chat application meant for companies
+and similar groups ranging in size from 15 to more than a thousand
+users. It features real-time notifications, message persistence and
+search, public group conversations (*streams*), invite-only streams,
+private one-on-one and group conversations, inline image previews,
+team presence/a buddy list, a rich API, and several integrations with
+other services. The maintainer team aims to support users who connect
+to Zulip using dedicated iOS, Android, Linux, Windows, and Mac OS X
+clients, as well as people using modern web browsers or dedicated
+Zulip API clients.
+
+A server can host multiple Zulip *realms* (organizations) at the same
+domain, each of which is a private chamber with its own users,
+streams, customizations, and so on. This means that one person might
+be a user of multiple Zulip realms. The administrators of a realm can
+choose whether to allow anyone to register an account and join, or
+only allow people who have been invited, or restrict registrations to
+members of particular groups (using email domain names or corporate
+single-sign-on login for verification).
+
+
+Components
+==========
+
+
+Tornado and Django
+------------------
+
+Django is the main web application server; Tornado runs the
+server-to-client real-time push system. The app servers are configured
+by the Supervisor configuration (which explains how to start the
+server processes) and the nginx configuration (which explains which
+HTTP requests get sent to which app server).
+
+Tornado is an asynchronous server, so it can scale really well, but
+it's a real pain to write code for it. So Tornado gets only the parts
+that really need to scale because they involve maintaining a
+persistent connection from every single user. The parts that are
+activate relatively rarely (e.g. when people are typing) go into
+Django.
```

Comment:
We should mention that we attempt to as much as possible avoid doing cache or database queries inside the Tornado code paths, since those blocking requests carry a very high performance penalty for a single-threaded, asynchronous server.